### PR TITLE
Cherry Pick to 1.1: Preventing duplicate route entries

### PR DESCRIPTION
**Pull request by mbanikazemi**
Addresses issue #13267
Adds unit tests

### CODE REVIEW

**Comment by mbanikazemi** [comment]
This is PR #13431 being applied to release-1.1 as requested. @frankbu

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -947,3 +947,12 @@ func (ps *PushContext) initAuthorizationPolicies(env *Environment) error {
 	}
 	return nil
 }
+
+// AddVirtualServiceForTesting adds a virtual service to the push context.
+// It is to be used for TESTING ONLY.
+func (ps *PushContext) AddVirtualServiceForTesting(config *Config) {
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go]: This is a little suspicious. Not familiar enough with Go though - is this the right way to do this?

Comment (howardjohn) [reply pilot/pkg/model/push_context.go]: Nevermind I see this was already approved for master

**Comment by linsun** [comment]
/approve from release perspective

@duderino pls merge if you agree, change is minor mostly adds more testings.

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 2]
```diff
@@ -163,28 +164,40 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(env *m
...
-
 		virtualHosts := make([]route.VirtualHost, 0, len(virtualHostWrapper.VirtualServiceHosts)+len(virtualHostWrapper.Services))
 		for _, host := range virtualHostWrapper.VirtualServiceHosts {
-			virtualHosts = append(virtualHosts, route.VirtualHost{
-				Name:    fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port),
-				Domains: []string{host, fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port)},
-				Routes:  virtualHostWrapper.Routes,
-			})
+			name = fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port)
+			if _, found := uniques[name]; !found {
```
Comment (duderino) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: @costinm @rshriram, philosophical question here:  how hard should Pilot try to mirror Envoy's validation vs. delegating to Envoy and relaying the decision?

Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/httproute.go]: Well we don’t have a feedback loop. That’s about the crux of it. So we try really hard to validate and automatically figure out stuff

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 3]
```diff
@@ -163,28 +164,40 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(env *m
...
+				uniques[name] = struct{}{}
+				virtualHosts = append(virtualHosts, route.VirtualHost{
+					Name:    name,
+					Domains: []string{host, fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port)},
+					Routes:  virtualHostWrapper.Routes,
+				})
+			} else {
+				log.Warnf("httproute(buildSidecarOutboundHTTPRouteConfig): Duplicate route entry %v. Dropping.", name)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute.go]: Maybe we should add a metric  like https://github.com/istio/istio/blob/9b6d31b74d1c0cc9358cc82d395b53f71393326b/pilot/pkg/model/gateway.go#L48

**Comment by duderino** [comment]
Closing this in favor of https://github.com/istio/istio/pull/14219 because it has all these changes plus more tests

**Pull request closed without merging**